// ngx-ui-loader/esm5/lib/utils/interfaces.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/**
 * @record
 */
export function Time() { }
if (false) {
    /** @type {?|undefined} */
    Time.prototype.delay;
    /** @type {?|undefined} */
    Time.prototype.maxTime;
    /** @type {?|undefined} */
    Time.prototype.minTime;
}
/**
 * @record
 */
export function Task() { }
if (false) {
    /** @type {?|undefined} */
    Task.prototype.delayTimer;
    /** @type {?|undefined} */
    Task.prototype.isDelayed;
    /** @type {?} */
    Task.prototype.isForeground;
    /** @type {?|undefined} */
    Task.prototype.isOtherRunning;
    /** @type {?|undefined} */
    Task.prototype.maxTimer;
    /** @type {?|undefined} */
    Task.prototype.minTimer;
    /** @type {?|undefined} */
    Task.prototype.startAt;
    /** @type {?} */
    Task.prototype.taskId;
}
/**
 * @record
 */
export function Tasks() { }
/**
 * @record
 */
export function Loader() { }
if (false) {
    /** @type {?} */
    Loader.prototype.loaderId;
    /** @type {?} */
    Loader.prototype.tasks;
    /** @type {?|undefined} */
    Loader.prototype.isMaster;
    /** @type {?} */
    Loader.prototype.isBound;
}
/**
 * @record
 */
export function Loaders() { }
/**
 * @record
 */
export function ShowEvent() { }
if (false) {
    /** @type {?} */
    ShowEvent.prototype.loaderId;
    /** @type {?} */
    ShowEvent.prototype.isShow;
}
/**
 * @record
 */
export function Config() { }
if (false) {
    /** @type {?|undefined} */
    Config.prototype.exclude;
    /** @type {?|undefined} */
    Config.prototype.excludeRegexp;
    /** @type {?|undefined} */
    Config.prototype.loaderId;
    /** @type {?|undefined} */
    Config.prototype.showForeground;
}
/**
 * @record
 */
export function Exclude() { }
if (false) {
    /** @type {?} */
    Exclude.prototype.strs;
    /** @type {?} */
    Exclude.prototype.regExps;
}
/**
 * @record
 */
export function NgxUiLoaderConfig() { }
if (false) {
    /** @type {?|undefined} */
    NgxUiLoaderConfig.prototype.bgsColor;
    /** @type {?|undefined} */
    NgxUiLoaderConfig.prototype.bgsOpacity;
    /** @type {?|undefined} */
    NgxUiLoaderConfig.prototype.bgsPosition;
    /** @type {?|undefined} */
    NgxUiLoaderConfig.prototype.bgsSize;
    /** @type {?|undefined} */
    NgxUiLoaderConfig.prototype.blur;
    /** @type {?|undefined} */
    NgxUiLoaderConfig.prototype.bgsType;
    /** @type {?|undefined} */
    NgxUiLoaderConfig.prototype.fgsColor;
    /** @type {?|undefined} */
    NgxUiLoaderConfig.prototype.fgsPosition;
    /** @type {?|undefined} */
    NgxUiLoaderConfig.prototype.fgsSize;
    /** @type {?|undefined} */
    NgxUiLoaderConfig.prototype.fgsType;
    /** @type {?|undefined} */
    NgxUiLoaderConfig.prototype.gap;
    /** @type {?|undefined} */
    NgxUiLoaderConfig.prototype.logoPosition;
    /** @type {?|undefined} */
    NgxUiLoaderConfig.prototype.logoSize;
    /** @type {?|undefined} */
    NgxUiLoaderConfig.prototype.logoUrl;
    /** @type {?|undefined} */
    NgxUiLoaderConfig.prototype.masterLoaderId;
    /** @type {?|undefined} */
    NgxUiLoaderConfig.prototype.overlayBorderRadius;
    /** @type {?|undefined} */
    NgxUiLoaderConfig.prototype.overlayColor;
    /** @type {?|undefined} */
    NgxUiLoaderConfig.prototype.pbColor;
    /** @type {?|undefined} */
    NgxUiLoaderConfig.prototype.pbDirection;
    /** @type {?|undefined} */
    NgxUiLoaderConfig.prototype.pbThickness;
    /** @type {?|undefined} */
    NgxUiLoaderConfig.prototype.hasProgressBar;
    /** @type {?|undefined} */
    NgxUiLoaderConfig.prototype.text;
    /** @type {?|undefined} */
    NgxUiLoaderConfig.prototype.textColor;
    /** @type {?|undefined} */
    NgxUiLoaderConfig.prototype.textPosition;
}
/**
 * @record
 */
export function NgxUiLoaderHttpConfig() { }
/**
 * @record
 */
export function NgxUiLoaderRouterConfig() { }
